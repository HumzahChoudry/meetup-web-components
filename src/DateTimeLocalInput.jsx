import React from 'react';
import cx from 'classnames';

/**
* @module DateTimeLocalInput
*/
class DateTimeLocalInput extends React.Component {

	constructor(props) {
		super(props);
		this.processInput = this.processInput.bind(this);
	}

	processInput(e) {
		this.onChange({ datetime: e.target.value });
	}

	render() {
		const {
			id,
			value,
			className,
			...other
		} = this.props;

		const classNames = cx(
			'dateTimePicker',
			className
		);

		return (
			<input
				id={id}
				type='datetime-local'
				value={value}
				className={classNames}
				onChange={this.processInput}
				{...other} />
		);

	}
}

DateTimeLocalInput.propTypes = {};

export default DateTimeLocalInput;

