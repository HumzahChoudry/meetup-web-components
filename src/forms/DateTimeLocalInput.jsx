import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

/**
 * @module DateTimeLocalInput
 * @description renders html5 datetime-local input
*/
class DateTimeLocalInput extends React.Component {

	constructor(props) {
		super(props);
		// this.state = {
		// 	value: this.props.value ? (new Date(this.props.value).toISOString()).split('.')[0] : ''
		// 	// example: 2017-02-18T00:00:00
		// 	// leaving off milliseconds
		// 	// datetime local wont set value with milliseconds
		// };
		this.onChange = this.onChange.bind(this);
	}

	/**
	* @function onChange
	* @description sets state and calls a callback with the value
	* (callback used in wrapping components like datetimepicker)
	* @param e Event the change event
	*/
	onChange(e) {
		this.props.onChange && this.props.onChange(e.target.value);
		this.props.onChangeCallback && this.props.onChangeCallback(e.target.value);
	}

	render() {
		const {
			id,
			label,
			className,
			required,
			value,
			error,
			onChangeCallback,	// eslint-disable-line no-unused-vars
			onChange, 			// eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'input--dateTimeLocal',
			className
		);

		const labelClassNames = cx({required});
		return (
			<div>
				<label htmlFor={id} className={labelClassNames}>{label}</label>
				<input
					id={id}
					type='datetime-local'
					value={value}
					className={classNames}
					onChange={this.onChange}
					required={required}
					{...other}
				/>
				{ error && <p className='text--error'>{error}</p> }
			</div>
		);

	}
}

DateTimeLocalInput.propTypes = {
	id: PropTypes.string,
	label: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element,
	]),
	onChangeCallback: PropTypes.func
};

export default DateTimeLocalInput;

