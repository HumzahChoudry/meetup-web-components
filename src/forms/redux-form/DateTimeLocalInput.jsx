import React from 'react';
import DateTimeLocalInput from '../DateTimeLocalInput';

/**
 * This component wraps the standard DateTimeLocalInput
 * component for use with redux-form.
 * It deconstructs props that redux-form sets and re-sets them on DateTimeLocalInput
 *.
 * @param {object} props React component props
 * @return {React.Component} DateTimeLocalInput
 *
 */
const ReduxFormDateTimeLocalInput = props => {
	const { meta, input, ...other } = props;

	return <DateTimeLocalInput {...input} error={meta.error} {...other} />;
};

ReduxFormDateTimeLocalInput.displayName = 'ReduxFormDateTimeLocalInput';

export default ReduxFormDateTimeLocalInput;
