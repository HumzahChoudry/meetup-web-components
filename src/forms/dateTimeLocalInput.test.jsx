import React from 'react';
import { shallow, mount } from 'enzyme';
import DateTimeLocalInput from './DateTimeLocalInput';

describe('DateTimeLocal input', function() {
	let component,
		inputWrapper,
		handleChangeSpy;

	const value = '2000-01-01T00:01',
		newValue = '2017-06-01T08:30',
		min = '1999-12-31T23:55',
		max = '2017-06-30T16:30';

	const onChangePropMock = jest.fn();

	const props = {
		name: 'datetime',
		value,
		min,
		max,
		required: true,
		onChange:onChangePropMock,
	};

	beforeEach(() => {
		handleChangeSpy = jest.spyOn(DateTimeLocalInput.prototype, 'handleChange');
		component = shallow(<DateTimeLocalInput {...props}/>);
		inputWrapper = component.find('input');
	});

	afterEach(() => {
		component = null;
		inputWrapper = null;
		handleChangeSpy.mockClear();
	});

	it('should render an input with type time and expected attrs', function() {
		expect(component).toMatchSnapshot();
	});

	it('if controlled component without paren, should update its value and state on input change', function() {
		const propsWithoutValue = {
			name: 'datetime',
			defaultValue: '2000-01-01T00:01',
			min,
			max,
			required: true
		};
		const eventObj = { target: { value: newValue } };

		component = mount(<DateTimeLocalInput {...propsWithoutValue} />);
		inputWrapper = component.find('input');

		inputWrapper.simulate('change', eventObj);
		expect(component.state('value')).toEqual(newValue);
		expect(component.find('input').getDOMNode().getAttribute('value')).toEqual(newValue);
		// here is where I would like to confirm the value of the actual input to make sure 
		// the value from state went through
	});

	it('calls internal onChange when value changed', function() {
		const eventObj = { target: { value: newValue } };
		inputWrapper.simulate('change', eventObj);
		expect(handleChangeSpy).toHaveBeenCalledWith(eventObj);
	});

	it('calls onChange prop if one is provided, as with redux-form', function() {
		component.instance().handleChange({ target: { value: newValue } });
		expect(onChangePropMock).toHaveBeenCalledWith(newValue);
	});
});
