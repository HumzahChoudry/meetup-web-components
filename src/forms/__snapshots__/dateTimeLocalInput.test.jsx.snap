// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimeLocal input should render an input with type time and expected attrs 1`] = `
<div>
  <label
    className="required"
  />
  <input
    className="input--dateTimeLocal"
    max="3002-01-01T04:59:59.000Z"
    min="1999-12-30T05:00:00.000Z"
    name="datetime"
    onChange={[Function]}
    required={true}
    type="datetime-local"
    value="2000-01-01T04:59:59.000Z"
  />
</div>
`;
